### PR TITLE
feat: all properties have explicit defined (#23 fixes)

### PR DESCRIPTION
**Pull request by vic1707**
Since the original PR (#23) seems dead I took the liberty to copy the code and fix (I hope) all the requested changes. 

At time of writting I think I just have to split the rule in two as asked in one of the comments.

### CODE REVIEW

**Comment by vic1707** [comment]
The rules should be working in the current state, but the "non conflicting operators" one can be largely improved and needs to get the lists of incompatible decorators.
@darraghoriordan  do you know if the "typeCheckingDecorator" are all mutually exclusive (like setting `IsString` and `IsBoolean` are obviously incompatible)? If so it would drastically simplify the rule.

**Comment by vic1707** [comment]
The rules should be working in the current state, but the "non conflicting operators" one can be largely improved and needs to get the lists of incompatible decorators.
@darraghoriordan  do you know if the "typeCheckingDecorator" are all mutually exclusive (like setting `IsString` and `IsBoolean` are obviously incompatible)? If so it would drastically simplify the rule.

**Comment by darraghoriordan** [comment]
hey!

I had a look at this and the tests didn't pass on my machine. I downloaded it, `yarn && yarn build && yarn test`

**Comment by vic1707** [comment]
tests are now fixed @darraghoriordan.
Before considering the merge I'd still need some clues about the [non conflicting operators](https://github.com/darraghoriordan/eslint-plugin-nestjs-typed/pull/27#issuecomment-1150433379).
Also I'm searching for a way to get the quickfix to figure out the indentation of the code

**Comment by vic1707** [comment]
@darraghoriordan do we have an update on this?

**Comment by darraghoriordan** [comment]
> 

heya! I need a bit of time to understand this one better. no eta unfortunately. I'll try to get to it in Sept 2022 though! 👍

**Comment by vic1707** [comment]
okay 👍

**Comment by vic1707** [comment]
Hello @darraghoriordan 
I'm not trying to be a pain in the ass or whatever but it's been 1 month since you've said you'll check my PR. Did you got the time to look at it? Maybe you need some help to understand my code?

**Comment by darraghoriordan** [comment]
> Hello @darraghoriordan I'm not trying to be a pain in the ass or whatever but it's been 1 month since you've said you'll check my PR. Did you got the time to look at it? Maybe you need some help to understand my code?

hey! No sorry I just don't have the time. It's a large PR adding rules to the default set. 

It will take me hours to review and test it and I have too many other things that are higher priority at the moment.

**Comment by darraghoriordan** [comment]
I've resolved all the conflicts and made some changes. 

Can you merge https://github.com/darraghoriordan/eslint-plugin-nestjs-typed/tree/feature/all-properties-have-explicit-defined in here please

File: README.md [hunk 1]
```diff
@@ -29,10 +29,13 @@ Preventing bugs
 -   validate-nested-of-array-should-set-each
 -   validated-non-primitive-property-needs-type-decorator
 -   all-properties-are-whitelisted
+-   no-conflicting-decorators
 
 Security
 
 -   should-specify-forbid-unknown-values
+-   all-defined-property-must-be-decored
```
Comment (darraghoriordan) [inline README.md:37]: i wouldn't say these are security based

new rules should have an explanation also. These ones would need instructions on how to turn on also if not in the recommended set.

File: src/utils/classValidatorDecorators.ts [hunk 2]
```diff
@@ -1,100 +1,224 @@
...
+    "IsDefined",
+    "IsOptional",
 ];
+
+type ClassValidatorDecorators = typeof classValidatorDecorators[number];
+
+export const classValidatorDecoratorsConflicts: {
+    [decorator in ClassValidatorDecorators]: ClassValidatorDecorators[];
+} = {
+    Allow: [],
```
Comment (darraghoriordan) [inline src/utils/classValidatorDecorators.ts:119]: these lists would be really difficult to maintain.

File: src/rules/allOptionnalPropertyMustBeDecored/allOptionnalPropertyMustBeDecored.test.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+const ruleTester = new RuleTester({
+    parser: "@typescript-eslint/parser",
+    parserOptions: {
+        ecmaVersion: 2015,
+        tsconfigRootDir: tsRootDirectory,
+        project: "./tsconfig.json",
+    },
+});
+
+ruleTester.run("all-properties-have-explicit-defined", rule, {
```
Comment (darraghoriordan) [inline src/rules/allOptionnalPropertyMustBeDecored/allOptionnalPropertyMustBeDecored.test.ts:15]: this is checking the wrong rule i think

File: src/rules/allDefinedPropertyMustBeDecored/allDefinedPropertyMustBeDecored.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+export default rule;
+
+function hasValidDecorator(propertyDefinition: PropertyDefinition): boolean {
+    let hasValidDecorator = false;
+    if (!propertyDefinition.decorators) return hasValidDecorator;
+    for (const decorator of propertyDefinition.decorators) {
+        if (
+            decorator.expression.type === AST_NODE_TYPES.CallExpression &&
+            decorator.expression.callee.type === AST_NODE_TYPES.Identifier &&
+            (decorator.expression.callee.name === "IsDefined" ||
```
Comment (darraghoriordan) [inline src/rules/allDefinedPropertyMustBeDecored/allDefinedPropertyMustBeDecored.ts:52]: this logic doesn't work on models that are not DTOs. You'd have to check if there was a property with one of the class validator decorators, then assume this is a DTO, then run the rule for all the other properties. I get too many false postivies on a real nestjs project with this as it is

File: src/rules/allDefinedPropertyMustBeDecored/allDefinedPropertyMustBeDecored.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import {AST_NODE_TYPES, TSESLint} from "@typescript-eslint/utils";
+import {PropertyDefinition} from "@typescript-eslint/types/dist/generated/ast-spec";
+import {createRule} from "../../utils/createRule";
+import {typeCheckingDecorator} from "../../utils/classValidatorDecorators";
+
+const rule = createRule({
+    name: "all-defined-property-must-be-decored",
```
Comment (darraghoriordan) [inline src/rules/allDefinedPropertyMustBeDecored/allDefinedPropertyMustBeDecored.ts:7]: missing pluralisation convention that's on all other rules for "property"
spelling issue on "decored"

File: src/configs/recommended.ts [hunk 6]
```diff
@@ -28,5 +28,9 @@ export = {
         "@darraghor/nestjs-typed/validate-nested-of-array-should-set-each":
             "error",
         "@darraghor/nestjs-typed/all-properties-are-whitelisted": "error",
+        "@darraghor/nestjs-typed/no-conflicting-decorators": "error",
```
Comment (darraghoriordan) [inline src/configs/recommended.ts:31]: i'm not sure i'd add these to the default set

File: src/utils/classValidatorDecorators.ts [hunk 7]
```diff
@@ -1,100 +1,224 @@
-export const classValidatorDecorators = [
-    "IsBooleanString",
-    "IsPositive",
-    "IsLatLong",
-    "IsLongitude",
-    "IsLatitude",
-    "IsNegative",
+export const typeCheckingDecorator = [
```
Comment (darraghoriordan) [inline src/utils/classValidatorDecorators.ts:1]: i'm going to try to remove the need for these lists completely. They're too difficult to maintain

**Review by darraghoriordan — CHANGES_REQUESTED** [review]
Hey Vic, thanks for submitting the PR and for your patience. 

There are a few issues with the rule logic in it so i'm going close this. I've added comments to the parts so you can see for yourself but there's no need to fix them. I wont be merging this. 

I've had a look at iddans original PR and I've merged in the latest and changed a few things in there. i have to do some quick checks but I'll use that PR instead of this one to give you the same functionality as intended here.

**Comment by darraghoriordan** [comment]
see previous comments

**Comment by vic1707** [comment]
Okay, thx for the review and the release of the other pr

**Pull request closed without merging**